massa-final-state/src/final_state.rs

```diff
@@ -811,3 +829,460 @@ impl FinalState {
+            mip_store,
+            false,
+        )
+        .expect("Cannot init final state");
+
+        return (fstate, selector_receiver);
+    }
+
+    fn get_state_changes() -> StateChanges {
+        let mut state_changes = StateChanges::default();
+        let message = AsyncMessage::new(
+            Slot::new(1, 0),
+            0,
+            Address::from_str("AU12dG5xP1RDEB5ocdHkymNVvvSJmUL9BgHwCksDowqmGWxfpm93x").unwrap(),
+            Address::from_str("AU12htxRWiEm8jDJpJptr6cwEhWNcCSFWstN1MLSa96DDkVM9Y42G").unwrap(),
+            String::from("test"),
+            10000000,
+            Amount::from_str("1").unwrap(),
+            Amount::from_str("1").unwrap(),
+            Slot::new(2, 0),
+            Slot::new(3, 0),
+            vec![1, 2, 3, 4],
+            None,
+            None,
+        );
+        let mut async_pool_changes = AsyncPoolChanges::default();
+        async_pool_changes.0.insert(
+            message.compute_id(),
+            SetUpdateOrDelete::Set(message.clone()),
+        );
+        state_changes.async_pool_changes = async_pool_changes;
+
+        let amount = Amount::from_str("1").unwrap();
+        let bytecode = Bytecode(vec![1, 2, 3]);
+        let ledger_entry = LedgerEntryUpdate {
+            balance: SetOrKeep::Set(amount),
+            bytecode: SetOrKeep::Set(bytecode),
+            datastore: BTreeMap::default(),
+        };
+        let mut ledger_changes = LedgerChanges::default();
+        ledger_changes.0.insert(
+            Address::from_str("AU12dG5xP1RDEB5ocdHkymNVvvSJmUL9BgHwCksDowqmGWxfpm93x").unwrap(),
+            SetUpdateOrDelete::Update(ledger_entry),
+        );
+        state_changes.ledger_changes = ledger_changes;
+
+        let mut pos_changes = PoSChanges::default();
+        pos_changes.roll_changes = PreHashMap::default();
+        pos_changes.roll_changes.insert(
+            Address::from_str("AU12r1iM79EcS3sa4dmtUp28TiaPxK1weQcLsATcFoynPdukjdMqM").unwrap(),
+            0,
+        );
+        state_changes.pos_changes = pos_changes;
+
+        return state_changes;
+    }
+
+    #[test]
+    fn test_final_state_finalize() {
+        // 0- Create a final state
+        // 1- Attempt to finalize it with KO data
+        // 2- Attempt to finalize it with OK data
+
+        let (mut fstate, _selector_receiver) = get_final_state();
+        let initial_slot = Slot::new(0, 0);
+        assert_eq!(fstate.get_slot(), initial_slot);
+
+        let wrong_next_slot = Slot::new(1, 1);
+        let ok_next_slot = Slot::new(0, 1);
+        let changes = get_state_changes();
+
+        let res = fstate._finalize(wrong_next_slot, changes.clone());
+        assert!(res
+            .err()
+            .unwrap()
+            .to_string()
+            .contains("while the current slot is"));
+
+        assert_eq!(fstate.get_slot(), initial_slot);
+
+        let res = fstate._finalize(ok_next_slot, changes.clone());
+
+        assert!(res.is_err());
```

Comment:
Here, with "OK" data, we should still get a POS error? Seems a bit weird for me, maybe rename the test to `test_final_state_finalize_slot_check()` ?

Reply:
It's mainly just a readability issue for me. I understand why it errors out, but you should explain that in comments and/or rename the test so that it is clear that it should not finalize successfully.

Reply:
I've added a small comment so a future reader can easily understand why it returns an error :)